src/api/app/models/project.rb

```diff
@@ -16,6 +16,8 @@ class Project < ApplicationRecord
 
   before_destroy :cleanup_before_destroy
   after_destroy_commit :delete_on_backend
+  after_save :update_staging_workflow_on_backend
```

Comment:
Not having to create another API endpoint to delete a project is nice, but I don't like that we now have code specific to the staging workflow in the project... This model is already humongous.

Reply:
I wrote it somewhere here, we should avoid to have in controller or models   conditions like ```if(@prj.is_a?(StagingProject) ...``` or leak subclasses methods to it's parent class.. I would prefer to keep everything separated until we have a clear view of what is coming ahead

Reply:
I agree with you two, I don't like it either as I mentioned in [this comment](https://github.com/openSUSE/open-build-service/pull/7158#issuecomment-472073436).

However, if we keep it independent, the callbacks won't be triggered if we delete the staging project from the [project show page](https://build.opensuse.org/project/show/TestLinux:Staging:A) or from the old API endpoint `/source/<project>`. I mean, as a regular project.

I wonder if it's better to have that ugly code or to keep it independent and, in consequence , to deal with:

- providing 2 API endpoints (returning an error if someone tries to delete a staging project with the old endpoint)
- changing the link in the UI depending on the class.

@dmarcoux  @vpereira :thinking:

Reply:
So let me risk being mob-murdered: How about redefine StagingProject not as subclass of Project, but as a ProjectRole. A Project can have one role - which is a polymorphic model and can attach it to e.g. staging workflows. And these role models have callsback as your staging project has.

Reply:
> How about redefine StagingProject not as subclass of Project, but as a ProjectRole.

We already have this. It's called `Project.kind`

Reply:
this just a string. I'm talking about a model that you can delegate to.

Reply:
hi @saraycp , For me, we should keep the callback in the StagingProject model, have its own endpoint as described in the deliverable and re-think it later. If necessary in the source controller we do something like:

````
 raise NoCookiesForYouError if @project.is_a?(StagingProject)
````

To avoid that people try to delete StagingProject through the false endpoint (which wont happen, people will use the  ```osc``` commands available)

IMO, This PRCDD (Pull Request Comment Driven Design) (TM) isn't the way to go. We should refine the features somewhere else.

Reply:
remember you also need this is_a? in the view, And as they are after_save callbacks, you obviously also need them in every project editing happening.

Reply:
Hey @coolo!

> remember you also need this is_a? in the view, And as they are after_save callbacks, you obviously also need them in every project editing happening.

Yes, maybe, I'm not sure if it is needed, but maybe it is better to do something small like that, than start to redesign things before even ship the staging workflow

Reply:
> start to redesign things before even ship the staging workflow

But this is exactly what we are doing here....

Reply:
> But this is exactly what we are doing here....

In the PR? absolutely, but it would be better if we don't...

Reply:
> but it would be better if we don't...

So where else are you going to do this?

Reply:
as we are doing now, in our existing process (daily scrum) :)